File: src/handlers/templates.rs
use crate::error::{ErrorType, MapProcErr, ProcessError, WithItem};
use crate::handlers::entries::process_entry;
use crate::handlers::frontmatter::load_frontmatter_data;
use crate::handlers::pages::page;
use crate::utils::kv_replace;
use crate::utils::ProcessResult;
use fancy_regex::Regex;
use once_cell::sync::Lazy;
use serde_json::Value;
use std::{collections::HashSet, fs, path::PathBuf, str};

static TEMPLATE_REGEX: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r#"(?<!<!--)<::Template\{([A-Z][A-Za-z_]*(:[A-Z][A-Za-z_]*)*)\}\s*\/>(?!.*?-->)"#)
        .expect("Regex failed to parse. This shouldn't happen.")
});

pub fn get_template(src: &PathBuf, name: &str, mut hist: HashSet<PathBuf>) -> ProcessResult {
    let mut errors: Vec<ProcessError> = Vec::new();
    let template_path = src
        .join("templates")
        .join(name.replace(":", "/"))
        .with_extension("template.html");

    let data_path = src
        .join("data")
        .join(name.replace(":", "/"))
        .with_extension("data.json");

    if !hist.insert(template_path.clone()) {
        return ProcessResult {
            output: String::new(),
            errors: vec![ProcessError {
                error_type: ErrorType::Circular,
                item: WithItem::Template,
                path: template_path,
                message: Some(format!("{:?}", hist)),
            }],
        };
    }

    let template = match fs::read_to_string(&template_path).map_proc_err(
        WithItem::Template,
        ErrorType::Io,
        &template_path,
        Some("Failed to read template file".to_string()),
    ) {
        Ok(content) => content,
        Err(e) => {
            return ProcessResult {
                output: String::new(),
                errors: vec![e],
            };
        }
    };

    if template.is_empty() {
        errors.push(ProcessError {
            error_type: ErrorType::Other,
            item: WithItem::Template,
            path: template_path.clone(),
            message: Some("Template file is empty".to_string()),
        });
        return ProcessResult {
            output: String::new(),
            errors,
        };
    }

    // Try to load from .toml + frontmatter first, fall back to .json
    let toml_path = src
        .join("data")
        .join(name.replace(":", "/"))
        .with_extension("data.toml");

    let v: Value = if toml_path.exists() {
        // Use frontmatter-based loading
        match load_frontmatter_data(src, name) {
            Ok((value, fm_errors)) => {
                errors.extend(fm_errors);
                value
            }
            Err(fm_errors) => {
                errors.extend(fm_errors);
                return ProcessResult {
                    output: String::new(),
                    errors,
                };
            }
        }
    } else {
        // Fall back to JSON loading
        let data = match fs::read_to_string(&data_path).map_proc_err(
            WithItem::Data,
            ErrorType::Io,
            &data_path,
            Some("Failed to read data file".to_string()),
        ) {
            Ok(content) => content,
            Err(e) => {
                return ProcessResult {
                    output: String::new(),
                    errors: vec![e],
                };
            }
        };

        if data.is_empty() {
            errors.push(ProcessError {
                error_type: ErrorType::Other,
                item: WithItem::Data,
                path: data_path.clone(),
                message: Some("Data file is empty".to_string()),
            });
            return ProcessResult {
                output: String::new(),
                errors,
            };
        }

        match serde_json::from_str(&data) {
            Ok(value) => value,
            Err(e) => {
                errors.push(ProcessError {
                    error_type: ErrorType::Syntax,
                    item: WithItem::Data,
                    path: data_path,
                    message: Some(format!("JSON decode error: {}", e)),
                });
                return ProcessResult {
                    output: String::new(),
                    errors,
                };
            }
        }
    };

    let items = match v.as_array() {
        Some(array) => array,
        None => {
            errors.push(ProcessError {
                error_type: ErrorType::Syntax,
                item: WithItem::Data,
                path: data_path,
                message: Some("JSON wasn't an array".to_string()),
            });
            return ProcessResult {
                output: String::new(),
                errors,
            };
        }
    };

    let mut contents = String::with_capacity(template.len() * items.len());

    for object in items {
        let obj = match object.as_object() {
            Some(obj) => obj,
            None => {
                errors.push(ProcessError {
                    error_type: ErrorType::Syntax,
                    item: WithItem::Data,
                    path: data_path.clone(),
                    message: Some("Invalid object in JSON".to_string()),
                });
                continue;
            }
        };

        let mut entry_path = String::new();
        let mut result_path = String::new();
        let mut is_entry = false;
        let mut kv: Vec<(&str, &str)> = Vec::with_capacity(obj.len());

        for (key, value) in obj {
            let val = match value.as_str() {
                Some(s) => s,
                None => {
                    errors.push(ProcessError {
                        error_type: ErrorType::Syntax,
                        item: WithItem::Data,
                        path: data_path.clone(),
                        message: Some(format!(
                            "Value for key '{}' couldn't be decoded to string",
                            key
                        )),
                    });
                    continue;
                }
            };

            match key.as_str() {
                "--entry-path" => {
                    entry_path = val.to_string();
                    is_entry = true;
                }
                "--result-path" => {
                    result_path = val.to_string();
                    is_entry = true;
                }
                _ => {}
            }
            kv.push((key, val));
        }

        let processed_template = kv_replace(kv.clone(), template.clone());
        contents.push_str(&processed_template);

        if is_entry {
            let entry_errs = process_entry(src, name, entry_path, result_path, kv);
            errors.extend(entry_errs);
        }
    }

    let page_res = page(src, contents, hist);
    errors.extend(page_res.errors);
    ProcessResult {
        output: page_res.output,
        errors,
    }
}

pub fn process_template(src: &PathBuf, input: String, hist: HashSet<PathBuf>) -> ProcessResult {
    let mut errors = Vec::new();
    let mut output = input;

    // Early return if no templates
    if !TEMPLATE_REGEX.is_match(&output).unwrap_or(false) {
        return ProcessResult { output, errors };
    }

    let mut replacements = Vec::new();

    for f in TEMPLATE_REGEX.find_iter(&output) {
        match f {
            Ok(found) => {
                let found_str = found.as_str();
                let template_name = match found_str
                    .trim()
                    .strip_prefix("<::Template{")
                    .and_then(|s| s.strip_suffix("/>"))
                    .map(|s| s.trim().trim_end_matches('}'))
                {
                    Some(name) if !name.is_empty() => name,
                    _ => {
                        errors.push(ProcessError {
                            error_type: ErrorType::Syntax,
                            item: WithItem::Template,
                            path: PathBuf::new(),
                            message: Some(format!(
                                "Malformed template tag: '{}'. Expected <::Template{{Name}} />",
                                found_str
                            )),
                        });
                        continue;
                    }
                };

                let result = get_template(src, template_name, hist.clone());
                errors.extend(result.errors);
                replacements.push((found_str.to_owned(), result.output));
            }
            Err(e) => {
                errors.push(ProcessError {
                    error_type: ErrorType::Other,
                    item: WithItem::Template,
                    path: PathBuf::new(),
                    message: Some(format!("Regex error while scanning for templates: {}", e)),
                });
            }
        }
    }

    for (old, new) in replacements.into_iter().rev() {
        output = output.replacen(&old, &new, 1);
    }

    ProcessResult { output, errors }
}

File: src/handlers/templates.cpp
#include "handlers/templates.hpp"
#include "handlers/entries.hpp"
#include "handlers/frontmatter.hpp"
#include "handlers/pages.hpp"
#include <fstream>
#include <nlohmann/json.hpp>
#include <regex>
#include <sstream>

namespace simple::handlers {

static const std::regex template_regex{
    R"(<\::Template\{([A-Z][A-Za-z_]*(:[A-Z][A-Za-z_]*)*)\}\s*\/>)"};

auto get_template(const std::filesystem::path &src, std::string_view name,
                  std::unordered_set<std::filesystem::path> hist)
    -> ProcessResult {

  std::vector<ProcessError> errors;

  std::string name_str{name};
  size_t pos = 0;
  while ((pos = name_str.find(':', pos)) != std::string::npos) {
    name_str.replace(pos, 1, "/");
  }

  auto template_path = src / "templates" / (name_str + ".template.html");
  auto data_path = src / "data" / (name_str + ".data.json");

  if (!hist.insert(template_path).second) {
    return ProcessResult{
        "",
        {ProcessError{ErrorType::Circular, WithItem::Template, template_path,
                      "Circular dependency detected"}}};
  }

  std::ifstream template_file{template_path};
  if (!template_file) {
    return ProcessResult{
        "",
        {ProcessError{ErrorType::Io, WithItem::Template, template_path,
                      "Failed to read template file"}}};
  }

  std::stringstream template_buffer;
  template_buffer << template_file.rdbuf();
  auto template_str = template_buffer.str();

  if (template_str.empty()) {
    return ProcessResult{"", errors};
  }

  auto toml_path = src / "data" / (name_str + ".data.toml");
  nlohmann::json v;

  if (std::filesystem::exists(toml_path)) {
    auto fm_result = load_frontmatter_data(src, name);
    if (fm_result) {
      auto [value, fm_errors] = std::move(*fm_result);
      v = std::move(value);
      errors.insert(errors.end(), fm_errors.begin(), fm_errors.end());
    } else {
      errors.insert(errors.end(), fm_result.error().begin(),
                    fm_result.error().end());
      return ProcessResult{"", errors};
    }
  } else {
    std::ifstream data_file{data_path};
    if (!data_file) {
      return ProcessResult{
          "",
          {ProcessError{ErrorType::Io, WithItem::Data, data_path,
                        "Failed to read data file"}}};
    }

    std::stringstream data_buffer;
    data_buffer << data_file.rdbuf();
    auto data_str = data_buffer.str();

    try {
      v = nlohmann::json::parse(data_str);
    } catch (const nlohmann::json::exception &e) {
      return ProcessResult{
          "",
          {ProcessError{ErrorType::Syntax, WithItem::Data, data_path,
                        std::format("JSON decode error: {}", e.what())}}};
    }
  }

  if (!v.is_array()) {
    return ProcessResult{"",
                         {ProcessError{ErrorType::Syntax, WithItem::Data,
                                       data_path, "JSON wasn't an array"}}};
  }

  std::string contents;
  contents.reserve(template_str.size() * v.size());

  for (const auto &object : v) {
    if (!object.is_object()) {
      errors.push_back(ProcessError{ErrorType::Syntax, WithItem::Data,
                                    data_path, "Invalid object in JSON"});
      continue;
    }

    std::string entry_path;
    std::string result_path;
    bool is_entry = false;
    std::vector<std::pair<std::string, std::string>> kv_storage;
    std::vector<std::pair<std::string_view, std::string_view>> kv;

    for (auto it = object.begin(); it != object.end(); ++it) {
      if (!it.value().is_string()) {
        errors.push_back(
            ProcessError{ErrorType::Syntax, WithItem::Data, data_path,
                         "JSON object value couldn't be decoded to string"});
        continue;
      }

      auto key = it.key();
      auto val = it.value().get<std::string>();

      if (key == "--entry-path") {
        entry_path = val;
        is_entry = true;
      } else if (key == "--result-path") {
        result_path = val;
        is_entry = true;
      }

      kv_storage.emplace_back(key, val);
    }

    for (const auto &[k, v] : kv_storage) {
      kv.emplace_back(k, v);
    }

    auto processed_template = kv_replace(kv, template_str);
    contents += processed_template;

    if (is_entry) {
      auto entry_errs = process_entry(src, name, std::move(entry_path),
                                      std::move(result_path), kv);
      errors.insert(errors.end(), entry_errs.begin(), entry_errs.end());
    }
  }

  auto page_res = page(src, std::move(contents), std::move(hist));
  errors.insert(errors.end(), page_res.errors.begin(), page_res.errors.end());

  return ProcessResult{std::move(page_res.output), std::move(errors)};
}

auto process_template(const std::filesystem::path &src, std::string input,
                      std::unordered_set<std::filesystem::path> hist)
    -> ProcessResult {

  std::vector<ProcessError> errors;
  auto output = std::move(input);

  if (!std::regex_search(output, template_regex)) {
    return ProcessResult{std::move(output), errors};
  }

  std::vector<std::pair<std::string, std::string>> replacements;
  auto begin =
      std::sregex_iterator(output.begin(), output.end(), template_regex);
  auto end = std::sregex_iterator();

  for (auto it = begin; it != end; ++it) {
    std::smatch match = *it;

    if (is_inside_comment(output, match.position())) {
      continue;
    }

    auto found_str = match.str();
    auto template_name = match[1].str();

    auto result = get_template(src, template_name, hist);
    errors.insert(errors.end(), result.errors.begin(), result.errors.end());
    replacements.emplace_back(found_str, std::move(result.output));
  }

  for (auto it = replacements.rbegin(); it != replacements.rend(); ++it) {
    const auto &[old_val, new_val] = *it;
    auto pos = output.find(old_val);
    if (pos != std::string::npos) {
      output.replace(pos, old_val.length(), new_val);
    }
  }

  return ProcessResult{std::move(output), std::move(errors)};
}

} // namespace simple::handlers
